From 687a4ddbf1869870deab1a7b5ddeebf8036aab9b Mon Sep 17 00:00:00 2001
From: Florian Westphal <fwestpha@redhat.com>
Date: Tue, 2 Sep 2014 10:15:01 -0400
Subject: [net] vxlan: Checksum fixes

Message-id: <1409652920-431-39-git-send-email-fwestpha@redhat.com>
Patchwork-id: 90152
O-Subject: [PATCH 38/57] [RHEL7.1 net] vxlan: Checksum fixes
Bugzilla: 1131999
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: Vlad Yasevich <vyasevic@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>
RH-Acked-by: Jiri Benc <jbenc@redhat.com>

BZ: 1131999

commit f79b064c15068176e3f6f67715aafd7fe183120c
Author: Tom Herbert <therbert@google.com>
Date:   Sat Jun 14 23:24:36 2014 -0700

    vxlan: Checksum fixes

    Call skb_pop_rcv_encapsulation and postpull_rcsum for the Ethernet
    header to work properly with checksum complete.

    Signed-off-by: Tom Herbert <therbert@google.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/vxlan.c b/drivers/net/vxlan.c
index c00de04..ac87803 100644
--- a/drivers/net/vxlan.c
+++ b/drivers/net/vxlan.c
@@ -1143,15 +1143,7 @@ static int vxlan_udp_encap_recv(struct sock *sk, struct sk_buff *skb)
  if (!vs)
   goto drop;
 
- /* If the NIC driver gave us an encapsulated packet
-  * with the encapsulation mark, the device checksummed it
-  * for us. Otherwise force the upper layers to verify it.
-  */
- if ((skb->ip_summed != CHECKSUM_UNNECESSARY && skb->ip_summed != CHECKSUM_PARTIAL) ||
-     !skb->encapsulation)
-  skb->ip_summed = CHECKSUM_NONE;
-
- skb->encapsulation = 0;
+ skb_pop_rcv_encapsulation(skb);
 
  vs->rcv(vs, skb, vxh->vx_vni);
  return 0;
@@ -1187,6 +1179,7 @@ static void vxlan_rcv(struct vxlan_sock *vs,
  remote_ip = &vxlan->default_dst.remote_ip;
  skb_reset_mac_header(skb);
  skb->protocol = eth_type_trans(skb, vxlan->dev);
+ skb_postpull_rcsum(skb, eth_hdr(skb), ETH_HLEN);
 
  /* Ignore packet loops (and multicast echo) */
  if (compare_ether_addr(eth_hdr(skb)->h_source,
-- 
1.7.1